pkg/wallet/account/account.go

```diff
@@ -0,0 +1,152 @@
+package account
+
+import (
+	"crypto/ed25519"
+	"crypto/rand"
+	"fmt"
+
+	"github.com/awnumar/memguard"
+	"github.com/btcsuite/btcutil/base58"
+	"github.com/massalabs/station-massa-wallet/pkg/crypto"
+	"github.com/massalabs/station-massa-wallet/pkg/types"
+	"github.com/massalabs/station-massa-wallet/pkg/types/object"
+)
+
+const (
+	AccountLastVersion = 1
+)
+
+type Account struct {
+	Version      *uint8                    `yaml:"Version"`
+	Nickname     string                    `yaml:"Nickname"`
+	Address      types.Address             `yaml:"Address"`
+	Salt         [16]byte                  `yaml:"Salt,flow"`
+	Nonce        [12]byte                  `yaml:"Nonce,flow"`
+	CipheredData types.EncryptedPrivateKey `yaml:"CipheredData,flow"`
+	PublicKey    types.PublicKey           `yaml:"PublicKey,flow"`
+}
+
+func New(password *memguard.LockedBuffer, nickname string) (*Account, error) {
+	version := uint8(AccountLastVersion)
+
+	publicKeyBytes, privateKeyBytes, err := ed25519.GenerateKey(nil)
+	if err != nil {
+		return nil, fmt.Errorf("generating ed25519 key-pair: %w", err)
+	}
+
+	var salt [16]byte
+	_, err = rand.Read(salt[:])
+	if err != nil {
+		return nil, fmt.Errorf("generating random salt: %w", err)
+	}
+
+	var nonce [12]byte
+	_, err = rand.Read(nonce[:])
+	if err != nil {
+		return nil, fmt.Errorf("generating random nonce: %w", err)
+	}
+
+	aeadCipher, secretKey, err := crypto.NewSecretCipher(password.Bytes(), salt[:])
+	password.Destroy()
+	if err != nil {
+		return nil, fmt.Errorf("creating secret cipher: %w", err)
+	}
+
+	// secretBuffer is the secret key in clear without the version.
+	secretBuffer := memguard.NewBufferFromBytes(privateKeyBytes.Seed())
+	encryptedSecret := crypto.SealSecret(aeadCipher, nonce[:], secretBuffer)
+	secretKey.Destroy()
+
+	publicKey := types.PublicKey{
+		Object: &object.Object{
+			Kind:    object.PublicKey,
+			Version: types.PublicKeyLastVersion,
+			Data:    publicKeyBytes,
+		},
+	}
+
+	address := types.NewAddressFromPublicKey(&publicKey)
+
+	cipheredData := types.EncryptedPrivateKey{
+		Object: &object.Object{
+			Kind:    object.EncryptedPrivateKey,
+			Version: types.EncryptedPrivateKeyLastVersion,
+			Data:    encryptedSecret,
+		},
+	}
+
+	return &Account{
+		Version:      &version,
+		Nickname:     nickname,
+		Address:      *address,
+		Salt:         salt,
+		Nonce:        nonce,
+		CipheredData: cipheredData,
+		PublicKey:    publicKey,
+	}, nil
+}
+
+func NewFromPrivateKey(password *memguard.LockedBuffer, nickname string, privateKeyText *memguard.LockedBuffer) (*Account, error) {
```

Comment:
there is probably a way to avoid code duplication between those to constructor functions